{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Arg1:point mutation vcf\n",
    "#Arg2:tumor bam file\n",
    "\n",
    "#for BWA\n",
    "#updates\n",
    "#180503: cigar loop update: error correction for deletion spanning the alt\n",
    "#180518: cigar loop multiple error correction;add var_loca_reverse;change output format; get rlength from read\n",
    "#180518: add informations: presence of clipping(soft of hard), INS, DEL in ref of var reads \n",
    "#180531: add informations: min, max location, refNM, varNM\n",
    "#180717: indel is not available now ( write NA at all columns)\n",
    "#180830: remove indel nt count from nm count\n",
    "#180903: support gzip input\n",
    "#181010: print NA when variant read count was zero.\n",
    "#181022: indel can be annotated\n",
    "#181101: unmapped read, duplicate read -> 'blank'\n",
    "#181114: correction for multiple nucleotides in mutect2 such as CTT > CT, CT>AG\n",
    "#181128: deletion length calculation error correction\n",
    "#181129: deletion and insertion variant read detection error correction\n",
    "#190221: count by readname\n",
    "\n",
    "#by Youk\n",
    "#190225: output_file name change (delete '.vcf')\n",
    "#fetch(pos1-1,pos1+1) add for softclipping\n",
    "\n",
    "import sys,pysam,gzip\n",
    "from numpy import median\n",
    "print(sys.argv[1])\n",
    "if sys.argv[1][-3:]=='.gz':\n",
    "    in_file=gzip.open(sys.argv[1])\n",
    "else:\n",
    "    in_file=open(sys.argv[1])\n",
    "bam_file=pysam.AlignmentFile(sys.argv[2],'rb') #bam file\n",
    "if sys.argv[1][-3:]=='.gz':\n",
    "    out_file=open(sys.argv[1][:-3]+'.readinfo','w')\n",
    "else:\n",
    "    out_file=open(sys.argv[1]+'.readinfo','w')\n",
    "\n",
    "\n",
    "\n",
    "in_line=in_file.readline().strip()\n",
    "while in_line:\n",
    "    if in_line[0:4]=='#CHR':\n",
    "        out_file.write(in_line+'\\ttemp_ref_readN\\ttemp_var_readN\\tref_medMQ\\tvar_medMQ\\tvar_LocaLeftMin;Med;Max\\tvar_LocaRightMin;Med;Max\\trefClip;pct;varClip;pct\\trefIns;pct;varIns;pct\\trefDel;pct;varDel;pct\\tref_medMismatch\\tvar_medMismatch\\n')\n",
    "    elif in_line[0]=='#':\n",
    "        out_file.write(in_line+'\\n')\n",
    "    else:\n",
    "        in_indi=in_line.split('\\t')\n",
    "        chr1=in_indi[0]\n",
    "        pos1=int(in_indi[1])\n",
    "        ref_nt=in_indi[3]\n",
    "        alt_nt=(in_indi[4].split(',')[0]).split('/')[0]\n",
    "        if len(ref_nt) == 1 and len(alt_nt)==1:\n",
    "            mttype='snv'\n",
    "        elif len(ref_nt) == len(alt_nt) and len(ref_nt) >1:\n",
    "            mttype='snv'\n",
    "            ref_nt=ref_nt[0]\n",
    "            alt_nt=alt_nt[0]\n",
    "        elif len(ref_nt) > len(alt_nt):\n",
    "            mttype='del'\n",
    "            cor_len=len(alt_nt)-1\n",
    "            alt_nt=alt_nt[0]\n",
    "            ref_nt=ref_nt[0:len(ref_nt)-cor_len]\n",
    "            del_len=len(ref_nt)-1\n",
    "        elif len(ref_nt) < len(alt_nt):\n",
    "            mttype='ins'\n",
    "            cor_len=len(ref_nt)-1\n",
    "            ref_nt=ref_nt[0]\n",
    "            alt_nt=alt_nt[0:len(alt_nt)-cor_len]\n",
    "            ins_len=len(alt_nt)-1\n",
    "            ins_seq=alt_nt[1:]\n",
    "        else:\n",
    "            print('ERROR: unknown mutation type. exiting')\n",
    "            print(in_line)\n",
    "            sys.exit(1)\n",
    "        var_mapq=[];ref_mapq=[];var_loca_lt=[];var_loca_rt=[];var_nm=[];ref_nm=[]\n",
    "        ref_n_list=[];var_n_list=[]\n",
    "        ref_i=0;ref_d=0;ref_c=0;var_i=0;var_d=0;var_c=0\n",
    "        for read in bam_file.fetch(chr1,pos1-1,pos1+1):\n",
    "            if read.is_unmapped == True or read.is_duplicate == True: continue\n",
    "            var_read='off'\n",
    "            est_dist=pos1-read.reference_start-1\n",
    "            rlength=read.infer_query_length()\n",
    "            if read.cigartuples==None:\n",
    "                continue\n",
    "                \n",
    "            if read.reference_start == pos1 and read.query_alignment_start == 0:\n",
    "                continue              \n",
    "\n",
    "            cigar_list=read.cigartuples\n",
    "            current_m=0;current_i=0;current_d=0;target_del_stat=0\n",
    "            for cigar in cigar_list:   #loop for calculate real distance \n",
    "                if cigar[0]==0 and (current_m + current_d)  <=est_dist:\n",
    "                    current_m=current_m+cigar[1]\n",
    "                elif cigar[0]==1 and (current_m + current_d) <=est_dist:\n",
    "                    current_i=current_i+cigar[1]\n",
    "                elif cigar[0]==2 and (current_m + current_d) <=est_dist:\n",
    "                    if current_m+current_d+cigar[1] > est_dist:\n",
    "                        target_del_stat=1\n",
    "                        break\n",
    "                    else:\n",
    "                        current_d=current_d+cigar[1]\n",
    "                elif current_m + current_d > est_dist:\n",
    "                    break\n",
    "                else:\n",
    "                    'blank'\n",
    "            if target_del_stat==1:\n",
    "                ref_n_list.append(read.query_name)\n",
    "            rel_dist=est_dist+current_i-current_d  # start with 0\n",
    "\n",
    "            cigar_i=0;cigar_d=0;cigar_s=0;cigar_h=0; current_m=0;cigar_md=0  # cigar_md: any deletion after matching\n",
    "            for cigar in cigar_list:   #loop for check presence of clipping, insertion, deletion\n",
    "                if cigar[0]==0:\n",
    "                    current_m += cigar[1]\n",
    "                elif cigar[0]==1:\n",
    "                    if mttype == 'ins' and pos1-read.reference_start==current_m+cigar_md and cigar[1]==ins_len and read.query_alignment_sequence[rel_dist+1:rel_dist+1+ins_len]==ins_seq:\n",
    "                        var_read='on'\n",
    "                    else:\n",
    "                        cigar_i=cigar_i+1\n",
    "                elif cigar[0]==2:\n",
    "                    if mttype == 'del' and pos1-read.reference_start==current_m+cigar_md and cigar[1] == del_len:\n",
    "                        var_read='on'\n",
    "                    else:\n",
    "                        cigar_d=cigar_d+1\n",
    "                    if current_m > 0:\n",
    "                        cigar_md += cigar[1]\n",
    "                elif cigar[0]==4:\n",
    "                    cigar_s=cigar_s+1\n",
    "                elif cigar[0]==5:\n",
    "                    cigar_h=cigar_h+1\n",
    "            if target_del_stat !=1:\n",
    "                if mttype== 'snv' and read.query_alignment_sequence[rel_dist]==alt_nt:   #var_read\n",
    "                    var_read='on'\n",
    "\n",
    "            if var_read == 'on':\n",
    "                var_n_list.append(read.query_name)\n",
    "                var_mapq.append(read.mapping_quality)\n",
    "                if read.has_tag('NM')==True:\n",
    "                    current_nm=read.get_cigar_stats()[0][10] - (read.get_cigar_stats()[0][1]+read.get_cigar_stats()[0][2])\n",
    "                    var_nm.append(current_nm)\n",
    "                if cigar_list[0][0]==0:\n",
    "                    dist=rel_dist\n",
    "                    var_loca_lt.append(dist)\n",
    "                    var_loca_rt.append(rlength-1-dist)\n",
    "                elif cigar_list[0][0]==4 or cigar_list[0][0]==5:\n",
    "                    dist=rel_dist+cigar_list[0][1]\n",
    "                    var_loca_lt.append(dist)\n",
    "                    var_loca_rt.append(rlength-1-dist)\n",
    "                else:\n",
    "                    'blank'\n",
    "\n",
    "                if cigar_i > 0:\n",
    "                    var_i = var_i+1\n",
    "                if cigar_d > 0:\n",
    "                    var_d = var_d+1\n",
    "                if cigar_s > 0 or cigar_h > 0:\n",
    "                    var_c = var_c+1\n",
    "            if var_read == 'off':  #ref_read\n",
    "                ref_n_list.append(read.query_name)\n",
    "                ref_mapq.append(read.mapping_quality)\n",
    "                if read.has_tag('NM')==True: \n",
    "                    current_nm=read.get_cigar_stats()[0][10] - (read.get_cigar_stats()[0][1]+read.get_cigar_stats()[0][2])\n",
    "                    ref_nm.append(current_nm)\n",
    "                if cigar_i > 0:\n",
    "                    ref_i = ref_i+1\n",
    "                if cigar_d > 0:\n",
    "                    ref_d = ref_d+1\n",
    "                if cigar_s > 0 or cigar_h > 0:\n",
    "                    ref_c = ref_c+1\n",
    "            else:\n",
    "                'blank'\n",
    "                \n",
    "                \n",
    "                \n",
    "        if len(set(ref_n_list) & set(var_n_list)) > 0:  # in special case with different NT among overlapped pair\n",
    "            var_n_list=list(set(var_n_list) - set(ref_n_list))\n",
    "            ref_n_list=list(set(ref_n_list) - set(var_n_list))\n",
    "        ref_n=len(list(set(ref_n_list)))\n",
    "        var_n=len(list(set(var_n_list)))\n",
    "        if len(ref_mapq)==0:\n",
    "            mr_mq='NA'\n",
    "        else:\n",
    "            mr_mq=str(median(ref_mapq))\n",
    "        if len(var_mapq)==0:\n",
    "            mv_mq='NA'\n",
    "        else:\n",
    "            mv_mq=str(median(var_mapq))\n",
    "        if len(var_loca_lt)==0:\n",
    "            vlocal='NA'; vlocar='NA'\n",
    "        else:\n",
    "            vlocalmin=str(min(var_loca_lt))\n",
    "            vlocalmed=str(median(var_loca_lt))\n",
    "            vlocalmax=str(max(var_loca_lt))\n",
    "            vlocarmin=str(min(var_loca_rt))\n",
    "            vlocarmed=str(median(var_loca_rt))\n",
    "            vlocarmax=str(max(var_loca_rt))\n",
    "            vlocal=';'.join([vlocalmin, vlocalmed, vlocalmax])\n",
    "            vlocar=';'.join([vlocarmin, vlocarmed, vlocarmax])\n",
    "        if ref_n==0:\n",
    "            ref_clip='NA;NA'\n",
    "            ref_ins='NA;NA'\n",
    "            ref_del='NA;NA'\n",
    "        else:\n",
    "            ref_clip=str(ref_c)+';'+str(round(ref_c*100/float(ref_n),2))\n",
    "            ref_ins=str(ref_i)+';'+str(round(ref_i*100/float(ref_n),2))\n",
    "            ref_del=str(ref_d)+';'+str(round(ref_d*100/float(ref_n),2))\n",
    "\n",
    "        if var_n==0:\n",
    "            var_clip='NA;NA'\n",
    "            var_ins='NA;NA'\n",
    "            var_del='NA;NA'\n",
    "        else:\n",
    "            var_clip=str(var_c)+';'+str(round(var_c*100/float(var_n),2))\n",
    "            var_ins=str(var_i)+';'+str(round(var_i*100/float(var_n),2))\n",
    "            var_del=str(var_d)+';'+str(round(var_d*100/float(var_n),2))\n",
    "        clip_info=ref_clip+';'+var_clip\n",
    "        ins_info=ref_ins+';'+var_ins\n",
    "        del_info=ref_del+';'+var_del\n",
    "\n",
    "        if len(ref_nm)==0:\n",
    "            mr_nm='NA'\n",
    "        else:\n",
    "            mr_nm=str(median(ref_nm))\n",
    "        if len(var_nm)==0:\n",
    "            mv_nm='NA'\n",
    "        else:\n",
    "            mv_nm=str(median(var_nm))\n",
    "\n",
    "        info_list=[str(ref_n), str(var_n), mr_mq, mv_mq, vlocal, vlocar, clip_info, ins_info, del_info, mr_nm, mv_nm ]\n",
    "        out_file.write(in_line+'\\t'+'\\t'.join(info_list)+'\\n')\n",
    "    in_line=in_file.readline().strip()\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
